### ISSUE TITLE
[syntax-errors] Remaining errors raised by the compiler

### ISSUE THREAD

**Issue by ntBre** [astral-sh/ruff#17412]
## Summary

A number of semantic or compile-time syntax errors have already been implemented
in #11934, but there are still many remaining ones. The purpose of this issue is
to track progress on this remainder. The initial source for most of these is
[this comment], which reports fuzzer results comparing Python's compiler and
ruff, but others may be included over time.

The next section lists the syntax errors that still need to be implemented, and
the sections after that include some more information on getting started with an
implementation.

### Errors mapping to current rules

These might be the easiest to start with because the existing rules have working
implementations and corresponding tests. However, they could possibly require
new context methods (see below), which could complicate things.

- `break` outside loop (F701)
- `continue` not properly in loop (F702)
- `return` with value in async generator (B901)
  - partial overlap with B901, which also applies to sync generators. This is
    only a syntax error in the async case
- `yield from` in async function (PLE1700)
- future feature is not defined (F407)
- import `*` only allowed at module level (F406)
- multiple starred expressions in assignment (F622)
- no binding for nonlocal (PLE0117)

### New errors

- generator expression must be parenthesized
``` python
"/".join(
    x for x in range(3), # this comma causes the issue
)
```
- can't use starred expression here
``` python
x = *value
```
- name is parameter and global
``` python
def f(a): global a
```
- name is parameter and nonlocal
``` python
def f(a): nonlocal a
```
- nonlocal declaration not allowed module level
``` python
nonlocal x
```
- not a chance (easter egg)
``` python
from __future__ import braces
```
- source code string cannot contain null bytes
```shell
printf '# \x00' > example.py
```
This mostly seems like an issue in comments

### Encoding issues

These syntax errors are all related to #6791 and should only be implemented if
we decide to respect these alternate encodings in general.

These can also be a bit tricky to put into a file, so most of these have shell
code snippets for generating the input rather than Python code directly.

- ascii codec can't decode byte `...` in position `...`: ordinal not in `range(128)`

```shell
printf '# -*- coding: ascii -*-\n\xc2' > example.py
```

- charmap codec can't decode byte ...

```shell
printf '# -*- coding: cp1252 -*-\n\x8d' > example.py
```

- utf7 codec can't decode byte 0xc3

```shell
printf '# -*- coding: utf-7 -*-\n\xc3' > example.py
```

This one reports an `E902` error, but only if it's enabled. Otherwise we only
print a warning about invalid UTF-8.

- encoding problem: `...` with BOM

``` shell
printf '\ufeff# -*- coding: ascii -*-' > example.py
```

I think any encoding other than UTF-8 with a BOM (`\uFEFF`) will cause this
issue.

- invalid character
- invalid non-printable character
  - both this and `invalid character` mostly seem to happen with the
    `iso-8859-1` encoding and a unicode character

- unknown encoding

```python
# coding: not-a-real-encoding
```

## Implementation

Issue #11934 contains links to each PR implementing a new semantic syntax error.
These can be used as examples of what a new implementation looks like, but this
section has a few more general notes.

All of the semantic syntax errors currently live in [`semantic_errors.rs`] and
correspond to a variant of the [`SemanticSyntaxErrorKind`] enum. The
[`SemanticSyntaxChecker`] is responsible for emitting these errors, but unlike
other AST visitors, it does not generally traverse the AST on its owns[^1]. Instead,
it relies on a parent visitor to call its `visit_stmt` and `visit_expr` methods
in the parent visitor's `visit_stmt` or `visit_expr` methods.

The [`SemanticSyntaxChecker`] also tracks very little of its own state. Instead,
it again defers to the parent visitor via the [`SemanticSyntaxContext`] trait.
Examples of methods on this trait include:
* `python_version` -- returns the target Python version for linting
* `in_async_context` -- returns whether or not async code like `await` or `async
  for` loops are allowed in the current scope

Typically the parent visitor, like the [`Checker`] from the `ruff_linter` crate,
will implement [`SemanticSyntaxContext`] and pass itself to any
[`SemanticSyntaxChecker`] method requiring a context.

Another important context method is `report_semantic_error`, which may require
special handling of new [`SemanticSyntaxErrorKind`] variants. For example, some
syntax errors emitted by CPython overlap with existing ruff rules (e.g.
[`await-outside-async`
(PLE1142)](https://docs.astral.sh/ruff/rules/await-outside-async/)). These need
to be transformed into normal ruff [`Diagnostic`]s instead of being reported
directly as syntax errors. This happens within the
[`Checker::report_semantic_error`] method.

## Testing

There are two main ways of testing new semantic errors. The first, and easier,
method is to write inline parser tests using `test_ok` and `test_err` comments,
for example:

https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_python_parser/src/semantic_errors.rs#L80-L95

These are automatically extracted from the source code by the parser's
[`generate_inline_tests`] function and then run like normal snapshot tests.
These work very well for simple errors that don't require detailed information
from the [`SemanticSyntaxContext`].

For errors that do require such contextual information tracked by the parent
visitor, a better alternative is to use integration tests, such as those found
in `linter.rs` in the `ruff_linter` crate:

https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_linter/src/linter.rs#L1015-L1025

https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_linter/src/linter.rs#L1064-L1066

The second of these can be reused for other semantic errors that correspond to
ruff rules, while something like the first example can be used for new
syntax-error-specific tests. In both cases, these integration tests will take
advantage of the existing context tracking done by the real [`Checker`] instead
of trying to duplicate that tracking in the parser's test
[`SemanticSyntaxCheckerVisitor`].

## Comparing to the fuzzer results

Again, the initial errors in this issue were extracted from [this comment] in
#7633 by following this procedure:
1. Download and unzip the fuzzer results from [here](https://github.com/qarmin/Automated-Fuzzer/releases/download/test/Cpython.no.parsable.parsable.by.ruff.zip)
2. Run the Python script below to get a partially-deduplicated list of errors
3. Double check these errors against ruff and CPython's parser and compiler

<details>
<summary>Python script for processing fuzzer results</summary>

```python
import argparse
import compileall
import contextlib
import dataclasses
import io
import os
import re
import subprocess
from collections import Counter

FILE = re.compile(r'File "([^"]+)", line (\d+)')
MESSAGE = re.compile(r"SyntaxError: (.*)")


@dataclasses.dataclass
class Error:
    file: str
    line_number: int
    message: str


# Ruff rules already re-implemented as semantic syntax errors, not just rules
# that match CPython syntax errors.
RULES = [
    "F404",
    "F704",
    "F706",
    "PLE0118",
    "PLE1142",
]


def run_ruff(ruff_bin, path):
    return subprocess.run(
        [
            ruff_bin,
            "check",
            "--no-cache",
            "--silent",
            "--preview",
            "--config",
            f"lint.select={RULES}",
            path,
        ]
    ).returncode


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--ruff-bin", default="ruff")

    args = parser.parse_args()

    with (
        contextlib.redirect_stdout(io.StringIO()) as out,
        contextlib.redirect_stderr(os.devnull),
    ):
        compileall.compile_dir(".", force=True, quiet=1)

    seen_messages = Counter()
    errors = list()

    file, line_number = None, None
    for line in out.getvalue().splitlines():
        if m := FILE.search(line):
            file, line_number = m.groups()
        if m := MESSAGE.search(line):
            message = m[1]
            # combine repetitive messages into single entries
            for prefix in [
                "unknown encoding",
                "invalid escape sequence",
                "'ascii' codec can't decode",
                "future feature",
                "invalid character",
                "invalid non-printable character",
                "no binding for nonlocal",
            ]:
                if message.startswith(prefix):
                    message = prefix
                    break
            if (
                message not in seen_messages
                and run_ruff(args.ruff_bin, file) == 0
            ):
                errors.append(Error(file, line_number, message))

            seen_messages[message] += 1

    for error in sorted(errors, key=lambda e: e.message):
        count = seen_messages[error.message]
        print(f"{error.file}:{error.line_number}: {error.message} ({count})")


if __name__ == "__main__":
    main()
```
</p>
</details>

[^1]: There are some local violations of this principle. For example, the
    [`ReboundComprehensionVisitor`] recursively visits the expressions in a
    comprehension looking for any expressions rebinding one of the
    comprehension's iteration variables. However, this is a highly-localized AST
    traversal and won't have the same kind of performance implications as an
    additional full AST traversal.

[this comment]: https://github.com/astral-sh/ruff/issues/7633#issuecomment-1740424031
[`semantic_errors.rs`]: https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_python_parser/src/semantic_errors.rs
[`SemanticSyntaxErrorKind`]: https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_python_parser/src/semantic_errors.rs#L891
[`SemanticSyntaxChecker`]: https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_python_parser/src/semantic_errors.rs#L19
[`ReboundComprehensionVisitor`]: https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_python_parser/src/semantic_errors.rs#L1291
[`SemanticSyntaxContext`]: https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_python_parser/src/semantic_errors.rs#L1586C11-L1586C32
[`Checker`]: https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_linter/src/checkers/ast/mod.rs#L182
[`Diagnostic`]: https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_diagnostics/src/diagnostic.rs#L22
[`Checker::report_semantic_error`]: https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_linter/src/checkers/ast/mod.rs#L572
[`generate_inline_tests`]: https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_python_parser/tests/generate_inline_tests.rs#L24
[`SemanticSyntaxCheckerVisitor`]: https://github.com/astral-sh/ruff/blob/014bb526f45afd654e5f42db92310e98b5068a54/crates/ruff_python_parser/tests/fixtures.rs#L470

**Comment by abhijeetbodas2001** [comment]
> generator expression must be parenthesized 

Regardless of whether there is a trailing comma or not, the exact same AST is generated by ruff for the code (only the ranges are different).
Given this, how do we detect the trailing comma during the semantic syntax error phase?
Should this rather be detected during parsing?
For example, the `ast` module does not give an AST at all for this:

```py
>>> import ast; ast.parse("max(i for i in range(3),)")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.11/ast.py", line 50, in parse
    return compile(source, filename, mode, flags,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<unknown>", line 1
    max(i for i in range(3),)
        ^^^^^^^^^^^^^^^^^^^
SyntaxError: Generator expression must be parenthesized
```

**Comment by ntBre** [comment]
Oh you're exactly right, thanks! The fact that `ast.parse` fails means this is not a semantic syntax error and should instead be caught by our parser. I'll open a separate issue for that.

**Comment by 11happy** [comment]
Hii , I am new to the astral community , I have tried to implement syntax error for `yield from` in async functions. 
Looking forward to learn & contribute.

Thank you

**Comment by 11happy** [comment]
I have also tried to implement `F406` (mport * only allowed at module leve) in PR linked above.
Thank you

**Comment by 11happy** [comment]
Hii @ntBre , I have tried to implement  `F622` multiple starred expression (linked above) , can you please take a look whenever you get a chance.
Thank you

**Comment by 11happy** [comment]
Hi @ntBre , I’ve implemented the new error case where a name is both a parameter and global. I think implementation is similar for the case where a name is both a parameter and nonlocal. Do you recommend adding that in the same PR, or should I open a separate one?
Whenever you get a chance, I’d appreciate your review. Thanks a lot for your time and guidance!

**Comment by ntBre** [comment]
Thank you! Let's start with `global` and save `nonlocal` for a separate PR. Both of these are a bit trickier than they look initially because `global` handling in ty is quite different from Ruff, and Ruff doesn't have much of the `nonlocal` support that ty has, so we may not even want to tackle that one immediately.

**Comment by 11happy** [comment]
Sure thing : ) , I understand.

**Comment by 11happy** [comment]
I have tried to implement F407 , have linked the PR above.

Thank you : )

**Comment by 11happy** [comment]
Hii @ntBre , I am also working on this break outside loop F701 , this would require some changes to the context, I have added a new `in_loop_context` , could you please take a loop & let me know if I am headed in right direction. https://github.com/astral-sh/ruff/pull/20556

Thank you :)

**Comment by 11happy** [comment]
Hii @ntBre , I have tried to implement the semantic syntax error where alternative patterns bind different names (linked above) . Whenever you get a chance, I’d appreciate your review. 

Thank you : )

**Comment by 11happy** [comment]
Hii @ntBre , now as we have `in_loop_context` method I have implemented(ported) `F702` as semantic syntax error linking PR (https://github.com/astral-sh/ruff/pull/20869) & have also worked on a new semantic syntax error linked above where annotated name cannot be global. Would appreciate your review .

Thank you : )

**Comment by 11happy** [comment]
Linking PR for PLE0117 https://github.com/astral-sh/ruff/pull/21032

Thank you : )

**Comment by 11happy** [comment]
Linking the follow up PR for name is parameter and global for ty https://github.com/astral-sh/ruff/pull/21312

Thank  you : )

**Comment by dhruvmanila** [comment]
This is another one that's easy enough to do:

```python
In [7]: class Foo2[T3 = int, T4]: ...
  Cell In[7], line 1
    class Foo2[T3 = int, T4]: ...
                         ^
SyntaxError: non-default type parameter 'T4' follows default type parameter
```

It's interesting that this isn't raised by the parser while a similar error involving function parameters is being raised by the parser:

```shell
❯ echo "class Foo[T1 = int, T2]: ..." | python -m ast -                                      
Module(
   body=[
      ClassDef(
         name='Foo',
         body=[
            Expr(
               value=Constant(value=Ellipsis))],
         type_params=[
            TypeVar(
               name='T1',
               default_value=Name(id='int', ctx=Load())),
            TypeVar(name='T2')])])

❯ echo "def foo(t1 = int, t2): ..." | python -m ast -
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/dhruv/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/ast.py", line 1865, in <module>
    main()
    ~~~~^^
  File "/Users/dhruv/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/ast.py", line 1861, in main
    tree = parse(source, name, args.mode, type_comments=args.no_type_comments)
  File "/Users/dhruv/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/ast.py", line 50, in parse
    return compile(source, filename, mode, flags,
                   _feature_version=feature_version, optimize=optimize)
  File "<stdin>", line 1
    def foo(t1 = int, t2): ...
                      ^^
SyntaxError: parameter without a default follows parameter with a default
```

**Comment by 11happy** [comment]
> This is another one that's easy enough to do:
> 
> In [7]: class Foo2[T3 = int, T4]: ...
>   Cell In[7], line 1
>     class Foo2[T3 = int, T4]: ...
>                          ^
> SyntaxError: non-default type parameter 'T4' follows default type parameter
> It's interesting that this isn't raised by the parser while a similar error involving function parameters is being raised by the parser:
> 
> ❯ echo "class Foo[T1 = int, T2]: ..." | python -m ast -                                      
> Module(
>    body=[
>       ClassDef(
>          name='Foo',
>          body=[
>             Expr(
>                value=Constant(value=Ellipsis))],
>          type_params=[
>             TypeVar(
>                name='T1',
>                default_value=Name(id='int', ctx=Load())),
>             TypeVar(name='T2')])])
> 
> ❯ echo "def foo(t1 = int, t2): ..." | python -m ast -
> Traceback (most recent call last):
>   File "<frozen runpy>", line 198, in _run_module_as_main
>   File "<frozen runpy>", line 88, in _run_code
>   File "/Users/dhruv/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/ast.py", line 1865, in <module>
>     main()
>     ~~~~^^
>   File "/Users/dhruv/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/ast.py", line 1861, in main
>     tree = parse(source, name, args.mode, type_comments=args.no_type_comments)
>   File "/Users/dhruv/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/ast.py", line 50, in parse
>     return compile(source, filename, mode, flags,
>                    _feature_version=feature_version, optimize=optimize)
>   File "<stdin>", line 1
>     def foo(t1 = int, t2): ...
>                       ^^
> SyntaxError: parameter without a default follows parameter with a default

I have linked the PR.

Thank you : )

**Comment by 11happy** [comment]
> This is another one that's easy enough to do:
> 
> In [7]: class Foo2[T3 = int, T4]: ...
>   Cell In[7], line 1
>     class Foo2[T3 = int, T4]: ...
>                          ^
> SyntaxError: non-default type parameter 'T4' follows default type parameter
> It's interesting that this isn't raised by the parser while a similar error involving function parameters is being raised by the parser:
> 
> ❯ echo "class Foo[T1 = int, T2]: ..." | python -m ast -                                      
> Module(
>    body=[
>       ClassDef(
>          name='Foo',
>          body=[
>             Expr(
>                value=Constant(value=Ellipsis))],
>          type_params=[
>             TypeVar(
>                name='T1',
>                default_value=Name(id='int', ctx=Load())),
>             TypeVar(name='T2')])])
> 
> ❯ echo "def foo(t1 = int, t2): ..." | python -m ast -
> Traceback (most recent call last):
>   File "<frozen runpy>", line 198, in _run_module_as_main
>   File "<frozen runpy>", line 88, in _run_code
>   File "/Users/dhruv/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/ast.py", line 1865, in <module>
>     main()
>     ~~~~^^
>   File "/Users/dhruv/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/ast.py", line 1861, in main
>     tree = parse(source, name, args.mode, type_comments=args.no_type_comments)
>   File "/Users/dhruv/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none/lib/python3.13/ast.py", line 50, in parse
>     return compile(source, filename, mode, flags,
>                    _feature_version=feature_version, optimize=optimize)
>   File "<stdin>", line 1
>     def foo(t1 = int, t2): ...
>                       ^^
> SyntaxError: parameter without a default follows parameter with a default

I have linked the PR for ty one as well.
Thank you : )

**Comment by Gambit-Checkmate** [comment]
Hi, I’d like to help with the remaining syntax-error cases. I’m planning to start with one small compiler case, add the corresponding fixture and snapshot, and keep the change limited to the behavior described here. Is there a particular syntax-error variant the team would prefer contributors to tackle first?
